pkg/query/simplify_caveat.go

```diff
@@ -0,0 +1,303 @@
+package query
+
+import (
+	"context"
+
+	"github.com/authzed/spicedb/internal/caveats"
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// mergeContexts combines relationship context from a caveat expression with query context
+// Query context takes precedence over relationship context
+func mergeContexts(expr *core.CaveatExpression, queryContext map[string]any) map[string]any {
+	fullContext := make(map[string]any)
+
+	// Start with the relationship context from the caveat expression
+	if expr.GetCaveat() != nil && expr.GetCaveat().Context != nil {
+		for k, v := range expr.GetCaveat().Context.AsMap() {
+			fullContext[k] = v
+		}
+	}
+
+	// Overlay with query-time context (takes precedence)
+	for k, v := range queryContext {
+		fullContext[k] = v
+	}
+
+	return fullContext
+}
+
+// mergeContextsForExpression recursively merges contexts for complex expressions
+// For AND/OR expressions, it collects all relationship contexts from all children
+func mergeContextsForExpression(expr *core.CaveatExpression, queryContext map[string]any) map[string]any {
+	fullContext := make(map[string]any)
+
+	// Collect all relationship contexts from the expression tree
+	collectRelationshipContexts(expr, fullContext)
+
+	// Overlay with query-time context (takes precedence)
+	for k, v := range queryContext {
+		fullContext[k] = v
+	}
+
+	return fullContext
+}
+
+// collectRelationshipContexts recursively collects relationship contexts from a caveat expression tree
+func collectRelationshipContexts(expr *core.CaveatExpression, contextMap map[string]any) {
+	if expr == nil {
+		return
+	}
+
+	// If this is a leaf caveat, collect its context
+	if expr.GetCaveat() != nil && expr.GetCaveat().Context != nil {
+		for k, v := range expr.GetCaveat().Context.AsMap() {
+			if _, exists := contextMap[k]; !exists {
+				contextMap[k] = v
+			}
+		}
+	}
+
+	// If this is an operation, recursively collect from children
+	if expr.GetOperation() != nil {
+		for _, child := range expr.GetOperation().Children {
+			collectRelationshipContexts(child, contextMap)
+		}
+	}
+}
+
+// SimplifyCaveatExpression simplifies a caveat expression by applying AND/OR logic:
+// - For AND: if a caveat evaluates to true, remove it from the expression
+// - For OR: if a caveat evaluates to true, the entire expression becomes true
+// Returns:
+//   - simplified: the simplified expression (nil if unconditionally true)
+//   - passes: true if passes unconditionally or conditionally, false if fails
+//   - error: any error that occurred during simplification
+func SimplifyCaveatExpression(
+	ctx context.Context,
+	runner *caveats.CaveatRunner,
```

Comment:
super tiny nitpick but your godoc says "if a caveat evaluates to..", but here you have a `runner`. I suggest renaming `runner->evaluator` for more consistency